index.bs

```diff
@@ -317,36 +317,79 @@ Datagrams are encrypted and congestion controlled.
 <pre class="idl">
 interface mixin DatagramTransport {
     readonly attribute unsigned short maxDatagramSize;
-    /* both streams are streams of Uint8Array objects */
-    readonly attribute ReadableStream datagramReadable;
-    readonly attribute WritableStream datagramWritable;
+    readonly attribute DuplexStream datagrams;
 };
 </pre>
 
 ## Attributes ##  {#datagram-transport-attributes}
 
 : <dfn for="DatagramTransport" attribute>maxDatagramSize</dfn>
 :: The maximum size data that may be passed to
-   {{DatagramTransport/datagramWritable}}.
+   {{DatagramTransport/datagrams}}' {{DuplexStream/writable}}.
 
-: <dfn for="DatagramTransport" attribute>datagramWritable</dfn>
-:: Sends datagrams that are written to the returned {{WritableStream}}.
+: <dfn for="DatagramTransport" attribute>datagrams</dfn>
+:: A single duplex stream for sending and receiving datagrams over this connection.
 
-   The getter steps for the `datagramWritable` attribute SHALL be:
-     1. Return the value of the {{[[SentDatagrams]]}} internal slot.
+   Each datagram received will be readable on the {{ReadableStream}} member.
 
-: <dfn for="DatagramTransport" attribute>datagramReadable</dfn>
-:: Return the value of the {{[[ReceivedDatagrams]]}} internal slot.
-
-   For each datagram received, insert it into {{[[ReceivedDatagrams]]}}. If too
-   many datagrams are queued because the stream is not being read quickly
-   enough, drop datagrams to avoid queueing. Implementations should drop older
+   If too many datagrams are queued because the stream is not being read quickly
+   enough, datagrams are dropped to avoid queueing. Implementations should drop older
    datagrams in favor of newer datagrams. The number of datagrams to queue
    should be kept small enough to avoid adding significant latency to packet
    delivery when the stream is being read slowly (due to the reader being slow)
    but large enough to avoid dropping packets when for the stream is not read
    for short periods of time (due to the reader being paused).
 
+   Datagrams written to the returned {{WritableStream}} member will be sent.
+
+   The getter steps for the `datagrams` attribute SHALL be:
+     1. Return the value of the {{[[Datagrams]]}} internal slot.
+
+## Procedures ## {#datagrams-transport-procedures}
+
+The <dfn>readDatagrams</dfn> algorithm is given a |transport| as parameter. It
+is defined by running the following steps:
+1. TODO
+1. [=ReadableStream/enqueue=] |data| in |transport|.`[[IncomingDatagrams]]`.
+
+The <dfn>writeDatagrams</dfn> algorithm is given a |transport| as parameter and
+|data| as input. It is defined by running the following steps:
+1. TODO
+1. Enqueue data for datagram transmission.
+1. Return [=a promise resolved with=] undefined.
+
+# `DuplexStream` Interface #  {#duplex-stream}
+
+A <dfn interface>DuplexStream</dfn> is a generic duplex stream.
```

Comment:
Isn't this too generic? Do you want other specs to use this interface?

We may want to add datagram specific attributes/methods to this interface - for example, `WebTransport.setDatagramWritableQueueExpirationDuration()` proposed at #179 can be `WebTransport.datagrams.setWritableQueueExpirationDuration()`. What do you think about renaming this to a more datagram specific name, such as `Datagrams` or `WebTransportDatagrams`?


cc: @ricea

Reply:
Yes it's generic. I was actually surprised not to find it in the streams spec which already defines the term [duplex stream](https://streams.spec.whatwg.org/#duplex-stream). Agree we should move or specialize it, perhaps including a more generic one, if one becomes available? cc @domenic

> ... can be `WebTransport.datagrams.setWritableQueueExpirationDuration()`

Though I happen to disagree with https://github.com/w3c/webtransport/pull/179, I agree in theory. E.g. we might want to move our existing
```diff
- transport.maxDatagramSize
+ transport.datagrams.maxSize
```

Reply:
Indeed, the reason one isn't defined is because you shouldn't have an object with _just_ readable/writable properties. You could put readable/writable properties directly on DatagramTransport, for example, and that would make DatagramTransport a conceptual duplex stream.

Reply:
For what it's worth, I think `DatagramFlow` would be a good name given the terminology used elsewhere right now.

Reply:
@vasilvv where is that used? I was thinking `DatagramDuplexStream` would be the most straightforward and in keeping with the surrounding names.